### ISSUE TITLE
Module Include implementation

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#6206]
Today, kernel names have two different forms "canonical" and "user" names. This is a bit cumbersome for a variety of reasons, in particular due to requiring essentially all clients to pick an ordering which they often do get wrong.

I understand from talking to developers (@maximedenes) that this dichotomy was introduced to better support inclusion of modules. What other strategies would be possible?

**Comment by ppedrot** [comment]
> I understand from talking to developers (@maximedenes) that this dichotomy was introduced to better support inclusion of modules. What other strategies would be possible?

Really? I thought it was to implement module aliases instead, which occurs from mere functor applications...

**Comment by ejgallego** [comment]
I think that module include also introduces aliases, right? But indeed, the real reason is to support aliases I think.

Now, what should be done about that?

**Comment by ejgallego** [comment]
So a possible idea after some brainstorming with @maximedenes would be to have the kernel store only the canonical name and aliases in the environment? WDYT @coq/core ?

**Comment by aspiwack** [comment]
In a way this canonical name thing is a rough implementation (approximation?) of a union-find structure. Question: could we store aliases in a union-find structure in the environment and use this to compare names to a stronger equivalence than delta?

I'm sure this is not as easy as it sounds, but throwing it out there, in case it's useful.

Also, I'd like to understand better what the consequence would be to simply remove the canonical name,

Conversely, I'm guessing that the problem with canonical names can't be only that it is a slightly confusing API (otherwise, it's almost certainly not worth any reimplementation cost). So let me encourage everyone to use this issue to catalogue them.

**Comment by ejgallego** [comment]
> I'm sure this is not as easy as it sounds, but throwing it out there, in case it's useful.

Indeed, the main idea to move forward is to make all kernel-level names "canonical" and provide a resolution step from user steps storing the aliases in the environment. It is work indeed.

> Also, I'd like to understand better what the consequence would be to simply remove the canonical name,

I think that has been tried and quite a few places fail, as expected from the missing equalities. If you don't use modules you may be Ok thou.

> Conversely, I'm guessing that the problem with canonical names can't be only that it is a slightly confusing API (otherwise, it's almost certainly not worth any reimplementation cost).

Well, the naming implementation [I think this issues is beyond an API issue] is not slightly but extremely confusing for the reasons stated on Gitter, and indeed I was told that critical bugs happened because of this. AFAICS, having such an implementation choice I can only see more and more problems popping up, when should consumers choose one equality or the other for names?

**Comment by aspiwack** [comment]
> I think that has been tried and quite a few places fail, as expected from the missing equalities. If you don't use modules you may be Ok thou.

But beyond backward compatibility? Some of us must have some idea of what kind of havoc this will cause. What kind of use-cases wouldn't be supported anymore, for instance.

> extremely confusing for the reasons stated on Gitter

Sure but: Gitter is not a source of data, it's a place of discussion. And I was there, and still don't really understand the problem.

>  I was told that critical bugs happened because of this.

Excellent! Let's document them here.

**Comment by ejgallego** [comment]
Oh, I think a change on this should not produce an observable change for the user. The idea is to move the kernel alias to an internalization one [maybe].

Regarding Gitter, I fully agree with you, that's why I tried to suggest @maximedenes and @ppedrot to follow up here.
